### ISSUE TITLE
508-defect-1: Dates are not read out within tables on responsive list view

### ISSUE THREAD

**Issue by joshkimux** [department-of-veterans-affairs/va.gov-team#50934]
### Point of contact

joshkimux

### Blocked by Apple bug
- [AX: Tables with display properties are skipped, report wrong col/row counts](https://bugs.webkit.org/show_bug.cgi?id=257458)

### Severity level

1, Launchblocker. Critical. Must be fixed before launch.

### Details

Dates within tables is skipped on a responsive list view. I hypothesize this is due to the table having unclear semantics around being a single header table (only the first row acts as a header) or a double header table (the first row _and_ first column both act as headers) where:
- the table is designed to _look_ like a single header table _but_
- the dates act like a double header as they are written with `th` (instead of `td`) and have scope attributes
- the design system table does not support double header tables within the responsive list view format

https://user-images.githubusercontent.com/14154792/208690942-344a92ba-342c-41cc-9725-4450daddc154.mov

[I started a slack thread in #accessibility-help](https://dsva.slack.com/archives/C8E985R32/p1671545950578269) to get more context and double check this assessment.

### Reproduction steps

1. Use Safari and VoiceOver on desktop
2. Log in as user 228 on staging and [view payments](https://staging.va.gov/va-payment-history/payments/)
3. Zoom in until responsive list view table appears
4. Using control + option + arrow keys, confirm that the date is not read out

### Proposed solution or next steps

**This issue can be immediately fixed if**
- [ ] dates are `td` elements, not `th`s
- [ ] dates no longer have scope attributes

https://user-images.githubusercontent.com/14154792/208692509-a4611aaa-99ca-4ab3-a1d9-1a6d8ea214d9.mov

_but_, it's unsure what the intended behavior actually is. If this is actually meant to be a double header table, more complex work will need to be explored with the design system team. I'll wait to hear more feedback from others in slack before providing a recommended solution.

### References, articles, or WCAG support

1. [WCAG WAI Tables](https://www.w3.org/WAI/tutorials/tables/)

### Type of issue

- [X] Screenreader
- [ ] Keyboard
- [ ] Focus
- [ ] Headings
- [ ] Color
- [ ] Zoom
- [ ] Semantics
- [ ] Axe-core
- [ ] Something else

## Acceptance Criteria

- [ ]

**Comment by jerekshoe** [comment]
For context
This is what the table looks like in non-responsive mode
<img width="1031" alt="Screenshot 2023-03-28 at 3 05 19 PM" src="https://user-images.githubusercontent.com/13838621/228354426-3a2ddeab-ab3b-494f-a35f-3c1059e7ea32.png">

**Comment by jacobworrell** [comment]
@joshkimux @jaredcunha we're finally taking a close look at this issue. It's a tricky one. We are considering transitioning from the React table component to the web component, but we're unsure how many of the same defects cited here are present in that web component. What we propose is that you guys do an audit of that web component to see to what extent it meets accessibility standards. If it is more suitable than this one, we can take a look at an implementation approach that swaps this component for the web component.

**Comment by HeatherWidmont** [comment]
@joshkimux can you please provide an update on the above comment?

**Comment by joshkimux** [comment]
@HeatherWidmont thank you for the nudge! Adding this to my list of things to tackle

**Comment by HeatherWidmont** [comment]
Great, thanks Josh! 

Possibly related to: https://app.zenhub.com/workspaces/benefits-team-1-6138d7b57a2631001a4b7562/issues/gh/department-of-veterans-affairs/va.gov-team/55927

**Comment by joshkimux** [comment]
> We are considering transitioning from the React table component to the web component, but we're unsure how many of the same defects cited here are present in that web component. What we propose is that you guys do an audit of that web component to see to what extent it meets accessibility standards. If it is more suitable than this one, we can take a look at an implementation approach that swaps this component for the web component.

[I've created a CAIA ticket to address this](https://github.com/department-of-veterans-affairs/va.gov-team/issues/58146) as it's on the topic of auditing a design system web component. Depending on bandwidth, I may personally take this on behalf of the CAIA team.

**Comment by HeatherWidmont** [comment]
At the moment we are waiting for a response from the CAIA team on the ticket commented above.

**Comment by HeatherWidmont** [comment]
@jaredcunha do you remember the goal of this ticket & know where it stands?

**Comment by HeatherWidmont** [comment]
Thanks @jaredcunha! We're putting this back into the Product Backlog to discuss during Backlog Refinement.

cc @jacobworrell

**Comment by jacobworrell** [comment]
@jaredcunha according to Apple Forums, Apple has fixed what was supposedly the root cause of this behavior. Would you mind testing this page with a screen reader again and seeing if the issue still occurs?

**Comment by HeatherWidmont** [comment]
Followed up with @jaredcunha in slack: https://coforma.slack.com/archives/C03TPPQ2KQ8/p1695930425785069

**Comment by HeatherWidmont** [comment]
@davidakennedy how do you recommend we proceed?

**Comment by davidakennedy** [comment]
Hey everyone,

I did a good bit of testing and research on this and want to share my findings:

**The bug appears to be fixed now.**

Here's a screen recording of me testing it:

https://github.com/department-of-veterans-affairs/va.gov-team/assets/1473618/6dac8113-59f0-45cf-9998-98af92dcd4ad

In the video, I navigate to one of the tables via a shortcut in VoiceOver, then navigate up and down the table. The dates are read out in all parts of the table, not just the one area shown in the video. A few notes:

- A lot has changed since this ticket was filed. Most notably, the design system moved to web components. The key cause of the bug, which was using display properties on different table elements get skipped, isn't present because the new web component doesn't do that. It uses ARIA to create fake table elements with `<span>` elements. I have some concerns about the long-term reliability of this method, since it may cause more bugs with assistive technology, but in this case, it helps us.
- The bug is also closed in Webkit's bug tracker, but I couldn't find out what version of Safari it would ship in.
- It looks like USWDS [fixed this in its table component as well](https://designsystem.digital.gov/components/table/). They removed the display properties in [3.6.1](https://github.com/uswds/uswds/releases/tag/v3.6.1).
- @eli-oat served as a double check for me here, also testing this and confirming my findings.
- I'm planning on talking to the design system team to find out more about why they needed to use non-semantic elements to create the table.

Recommendations: Close this bug as resolved. Work with the design system to ensure custom table components don't have any negative effects on assistive technology experience.
